### PR TITLE
feat: stream non-essential data

### LINKED ISSUE #7213: option to render layout before page is loaded
### Describe the problem

When building application-like websites, users expect instant feedback when changing pages, even if the page's content isn't ready yet. It's confusing to a user when they click a button and nothing happens, then a second later the page changes. Loading indicators help but are still not ideal.

### Describe the proposed solution

Provide an option that will allow a `+layout.svelte` file to be rendered before the `+page.svelte` is ready. For example, say clicking a button opens up a page that is a modal. We place the markup for the modal inside the `+layout.svelte` file and the content inside the `+page.svelte` file. As soon as the user clicks, we can render an empty modal. Then once the page is ready, the content will appear. While the page is loading, we can utilize the slot default as a loading page or skeleton UI.
```html
<script>
	// declare that this layout will be rendered to the DOM before the page is loaded
	export const optimistic = true;
</script>

<div class="modal">
	<slot>
		<p>skeleton ui or loading indicator</p>
	</slot>
</div>
```

### Alternatives considered

- **Using `#await` block so that surrounding markup is not dependent on data:** This destroys SSR and sveltekit's entire data model.
- **Having a dedicated `+loading.svelte` file:** Seems like unnecessary complexity when it can be done without it. If for some reason someone requires separate default and placeholder page contents they can always use the `$navigating` store to switch between them.

### Importance

would make my life easier

### Additional Information

- If the `+layout.svelte` also relies on data we can still render it as soon as its own data is ready.
- This probably shouldn't affect SSR in any way. When we render on the server, we probably want to wait until the entire page is loaded before responding to the client. If we didn't, it would defeat the purpose of SSR in the first place.
- As far as I know, this doesn't break any existing functionality.

### PR DESCRIPTION
**Pull request by dummdidumm**
Closes #7213 

This implements ~the proposed `defer` utility (inspired by React Router). At the moment this only works for client-side navigations. To implement this for SSR I see a couple of options. They all start with streaming the initial HTML which contains `_deferred_X` strings in place where the promises are that are still pending. The options revolve around how the chunks look like that come in later~ streaming for server load functions which you can utilize in the same way you can today with `+page.js` `load` functions by nesting the promise.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by dummdidumm** [comment]
@Rich-Harris brought up the nice idea of moving parts of this into `devalue`. `devalue` could take a custom serialize function, which we use to tell it to transform promises to something like `['Promise', 'id_of_the_promise']` - that way we also don't need to worry about our id strings clashing with anything the user might've written. It would also enable us to defer promises at _any_ level in the returned object, which is a nice QOL feature. Theoretically that makes the `defer` functions obsolete since you could move promises you want to defer into a nested object, but the ergonomics of that are not that nice IMO so I would be favor of keeping it. For v2 we could think about switching the logic around: top level promises are not auto-awaited, instead they function like if they were wrapped with `defer`.

**Comment by Rich-Harris** [comment]
PR to devalue: https://github.com/Rich-Harris/devalue/pull/58

**Comment by Rich-Harris** [comment]
I think it might be weird to introduce a `defer` function whose job is to _not_ transform its input (whereas normally the returned object is transformed).

I'd personally find this...

```js
export function load({ fetch }) {
  return {
    post: fetch(url).then(r => r.json()),
    lazy: {
      comments: fetch(url).then(r => r.json())
    }
  };
}
```

```svelte
<script>
  export let data;
</script>

<div>{@html data.post}</div>
{#await data.lazy.comments}
  <p>loading comments...</p>
{:then comments}
  <!-- TODO -->
{:then error
  <!-- TODO -->
{/await}
```

...more natural than this:

```js
import { defer } from '@sveltejs/kit';

export function load({ fetch }) {
  return defer({
    post: await fetch(url).then(r => r.json()),
    comments: fetch(url).then(r => r.json())
  });
}
```

Case in point: it's not until after I wrote that example that I realised it contains a waterfall.

If we do feel like awaiting top-level promises is confusing/a mistake and decide to change it in v2, then we could always add a config setting that opts you into v2 behaviour — we've done that for Svelte in the past. Then you'd just have to be careful about avoiding waterfalls:

```js
export function load({ fetch }) {
  return {
    comments: fetch(url).then(r => r.json()), // streamed, because no `await`
    post: await fetch(url).then(r => r.json()) // ready on render, because `await`
  };
}
```

**Comment by Rich-Harris** [comment]
Released `devalue@4.3.0` with custom types support

**Comment by dummdidumm** [comment]
Nice!
Note for monday: Keeping track of which resolved promise belongs to which node is super complicated. In the current design it would be needed to send along the final `uses` object. But since it's so complicated I think it's easier to send on final `{ type: uses, nodes: [{ url: 1, ..}, ..] }` chunk once everything else is completed.

**Comment by Rich-Harris** [comment]
I'm not sure I follow — why do we need to track that? Surely all that matters is the promise `id`?

**Comment by dummdidumm** [comment]
If we stream a promise we don't know if all the `uses` fields are final yet. After a promise is resolved someone could access `params`, which only then are markes as used (if not accessed before)

**Comment by Rich-Harris** [comment]
I think it would be totally fine if we said that you need to access those things before returning from `load` — I don't think I would have expected it to work differently as a user. Maybe we try shipping without that and see if it turns out to be an issue in practice?

**Comment by paoloricciuti** [comment]
I'm commenting on this because I've implemented a version of defer in user land. I would probably publish this library soon but I will be more than happy to deprecate it when this land stable in SK.

Currently actual streaming is not supported in the load function so I've had to work around this by using a dedicated endpoint with server sent events but I think I found a decent enough solution (would love some insight on this from some other maintainers, for the moment I've talked about this to Geoff and Antony.

So basically here's how my library works, you have to add a provided handle function in hooks.server.ts, than in your +page.server.ts you wrap the load function inside a provided defer function. Finally you access the data through a store (so not with export let data). And that's it.

- The handle function add two endpoints to your project, a GET and a POST: the GET endpoint uses server sent event to notify the client whenever one of the Promises on the server has resolved and the POST endpoint is used to trigger such event
- The defer function takes the return value of your load function and keep track of your Promises sticking them into an array.
- Once on the client the store create an EventStream and register a callback on the server sent event, it take the data returned from the load function and for every key in the promises array it create a new Promise saving the resolver and the rejecter in a Map
- Once the promise on the server resolves a post request is made to the POST endpoint managed by the handle function
- A new event is put out from the server sent event endpoint
- The callback registered in the store fires and the store resolves the correct Promise giving you back the data.

From a security standpoint the first time you call a load function it generates a secure cookie with a random UUID that uses to just get your events from the server.

Here's the link to the repo if you want to take a look:

https://github.com/paoloricciuti/sveltekit-defer

Currently this works fine with adapter node.

**Comment by Suyashtnt** [comment]
How will `defer` work if the client doesn't have JS enabled?

**Comment by paoloricciuti** [comment]
> How will `defer` work if the client doesn't have JS enabled?

I don't think it could ever work...you have to await a promise on the client for defer to have sense so no-js = no ability to await promises.

**Comment by Suyashtnt** [comment]
> > How will `defer` work if the client doesn't have JS enabled?
> 
> I don't think it could ever work...you have to await a promise on the client for defer to have sense so no-js = no ability to await promises. 

Maybe allow an option which makes defer only run on client side navigation. On SSR, it resolves on the server. If it's client side navigation, defer well, defers.

**Comment by Conduitry** [comment]
The problem there is that all SSR in Svelte is currently synchronous, and you can't synchronously tell whether a promise has already resolved. So you would need to write your component in a way that it can accept either direct data or a promise resolving to data. At that point, if you're already doing that, you may as well just also write your `load` function so that it checks whether you're on the browser and then either conditionally calls `defer` or `await`s the promise itself.

**Comment by david-plugge** [comment]
I'm pretty sure this would work:
```ts
function deferClient(data) {
  return ssr ? data : defer(data)
} 
```

**Comment by Rich-Harris** [comment]
To be clear, deferring will only apply to server loads, not universal loads. If you return an object containing a promise from a universal load function, its resolved value will be discarded and the promise will be created anew when the universal load function runs in the browser. (You can do this today, by the way — it's just how universal load functions work.)

So you wouldn't be detecting whether you're in the browser or not, you'd be detecting whether the server load function is running for SSR or for client-side navigation:

```js
// src/routes/whatever/+page.js
export function load({ isDataRequest }) {
  return {
    criticalData: getCriticalData(),
    lazyData: {
      whatever: isDataRequest ? getLazyData() : await getLazyData()
    }
  }
}
```

In your component, this would work whether the data was awaited or not:

```svelte
{#await data.lazyData.whatever}
  loading...
{:then whatever}
  {whatever}
{:catch error}
  {error.message}
{/await}
```

To be clear, I'm not necessarily suggesting you actually do this, just that you _can_.

**Comment by Suyashtnt** [comment]
> To be clear, deferring will only apply to server loads, not universal loads. If you return an object containing a promise from a universal load function, its resolved value will be discarded and the promise will be created anew when the universal load function runs in the browser. (You can do this today, by the way — it's just how universal load functions work.)
> 
> So you wouldn't be detecting whether you're in the browser or not, you'd be detecting whether the server load function is running for SSR or for client-side navigation:
> 
> ```js
> // src/routes/whatever/+page.js
> export function load({ isDataRequest }) {
>   return {
>     criticalData: getCriticalData(),
>     lazyData: {
>       whatever: isDataRequest ? getLazyData() : await getLazyData()
>     }
>   }
> }
> ```
> 
> In your component, this would work whether the data was awaited or not:
> 
> ```svelte
> {#await data.lazyData.whatever}
>   loading...
> {:then whatever}
>   {whatever}
> {:catch error}
>   {error.message}
> {/await}
> ```
> 
> To be clear, I'm not necessarily suggesting you actually do this, just that you _can_.

Just to make sure, this works on +server.page.ts right? I'm fetching the data via prisma, so I can't really use it on the client

**Comment by paoloricciuti** [comment]
> > To be clear, deferring will only apply to server loads, not universal loads. If you return an object containing a promise from a universal load function, its resolved value will be discarded and the promise will be created anew when the universal load function runs in the browser. (You can do this today, by the way — it's just how universal load functions work.)
> > So you wouldn't be detecting whether you're in the browser or not, you'd be detecting whether the server load function is running for SSR or for client-side navigation:
> > ```js
> > // src/routes/whatever/+page.js
> > export function load({ isDataRequest }) {
> >   return {
> >     criticalData: getCriticalData(),
> >     lazyData: {
> >       whatever: isDataRequest ? getLazyData() : await getLazyData()
> >     }
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In your component, this would work whether the data was awaited or not:
> > ```svelte
> > {#await data.lazyData.whatever}
> >   loading...
> > {:then whatever}
> >   {whatever}
> > {:catch error}
> >   {error.message}
> > {/await}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To be clear, I'm not necessarily suggesting you actually do this, just that you _can_.
> 
> Just to make sure, this works on +server.page.ts right? I'm fetching the data via prisma, so I can't really use it on the client

I might be wrong but what I think @Rich-Harris is saying is that from the universal load function (so not the `+page.server.ts`) you can actually return a promise already (but not as the top level of the returned object) but when it arrives on the client it simply get recreated from scratch (so it start loading on mount). This aligns with my experiments.

**Comment by Suyashtnt** [comment]
> > > To be clear, deferring will only apply to server loads, not universal loads. If you return an object containing a promise from a universal load function, its resolved value will be discarded and the promise will be created anew when the universal load function runs in the browser. (You can do this today, by the way — it's just how universal load functions work.)
> > > So you wouldn't be detecting whether you're in the browser or not, you'd be detecting whether the server load function is running for SSR or for client-side navigation:
> > > ```js
> > > // src/routes/whatever/+page.js
> > > export function load({ isDataRequest }) {
> > >   return {
> > >     criticalData: getCriticalData(),
> > >     lazyData: {
> > >       whatever: isDataRequest ? getLazyData() : await getLazyData()
> > >     }
> > >   }
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In your component, this would work whether the data was awaited or not:
> > > ```svelte
> > > {#await data.lazyData.whatever}
> > >   loading...
> > > {:then whatever}
> > >   {whatever}
> > > {:catch error}
> > >   {error.message}
> > > {/await}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > To be clear, I'm not necessarily suggesting you actually do this, just that you _can_.
> > 
> > 
> > Just to make sure, this works on +server.page.ts right? I'm fetching the data via prisma, so I can't really use it on the client
> 
> I might be wrong but what I think @Rich-Harris is saying is that from the universal load function (so not the `+page.server.ts`) you can actually return a promise already (but not as the top level of the returned object) but when it arrives on the client it simply get recreated from scratch (so it start loading on mount). This aligns with my experiments.

so with server-fetched data, Should I create an api route and call that in +page.ts?

**Comment by dummdidumm** [comment]
If this PR lands, you can use the same trick you can use in `+page.ts` files today in `+page.server.ts` files

**Comment by paoloricciuti** [comment]
> > > > To be clear, deferring will only apply to server loads, not universal loads. If you return an object containing a promise from a universal load function, its resolved value will be discarded and the promise will be created anew when the universal load function runs in the browser. (You can do this today, by the way — it's just how universal load functions work.)
> > > > So you wouldn't be detecting whether you're in the browser or not, you'd be detecting whether the server load function is running for SSR or for client-side navigation:
> > > > ```js
> > > > // src/routes/whatever/+page.js
> > > > export function load({ isDataRequest }) {
> > > >   return {
> > > >     criticalData: getCriticalData(),
> > > >     lazyData: {
> > > >       whatever: isDataRequest ? getLazyData() : await getLazyData()
> > > >     }
> > > >   }
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > In your component, this would work whether the data was awaited or not:
> > > > ```svelte
> > > > {#await data.lazyData.whatever}
> > > >   loading...
> > > > {:then whatever}
> > > >   {whatever}
> > > > {:catch error}
> > > >   {error.message}
> > > > {/await}
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > To be clear, I'm not necessarily suggesting you actually do this, just that you _can_.
> > > 
> > > 
> > > Just to make sure, this works on +server.page.ts right? I'm fetching the data via prisma, so I can't really use it on the client
> > 
> > 
> > I might be wrong but what I think @Rich-Harris is saying is that from the universal load function (so not the `+page.server.ts`) you can actually return a promise already (but not as the top level of the returned object) but when it arrives on the client it simply get recreated from scratch (so it start loading on mount). This aligns with my experiments.
> 
> so with server-fetched data, Should I create an api route and call that in +page.ts?

You can do that but as of now you still will not be able to avoid waterfalls (in the sense that it a promise from the server take 10s for example you will wait the whole 10s once it gets to the client.

**Comment by paoloricciuti** [comment]
> If this PR lands, you can use the same trick you can use in `+page.ts` files today in `+page.server.ts` files

@dummdidumm might ask you your opinion about the library I shared?

**Comment by dummdidumm** [comment]
Just skimmed the README, impressive work - though it's probably not needed anymore if this PR lands.

**Comment by paoloricciuti** [comment]
> Just skimmed the README, impressive work - though it's probably not needed anymore if this PR lands.

Yeah as I've said I'm more than happy to deprecate the library as soon as this land 😍

**Comment by paoloricciuti** [comment]
@dummdidumm I was reading the description again...may I ask you if the intended use of this is to actually defer the loading from the server without restarting the promise on the client (which should be the current behavior of the universal load function)? What I mean is if I fetch a slow endpoint without awaiting it, then await a faster endpoint and than return everything will the slow endpoint lose the progress made once it gets on the client or it will "stream" the promise resolution to the client?

**Comment by dummdidumm** [comment]
If you use `+page.js` it will "restart" the promise. If you use `+page.server.js`, it will stream in the promise result (no restart).

**Comment by paoloricciuti** [comment]
> If you use `+page.js` it will "restart" the promise. If you use `+page.server.js`, it will stream in the promise result (no restart).

Wonderful ☺️

File: documentation/docs/20-core-concepts/20-load.md
Comment (geoffrich) [inline documentation/docs/20-core-concepts/20-load.md:453]: This section duplicates some parts of the section above ("Promise unwrapping") - maybe we should merge them?

**Comment by Suyashtnt** [comment]
> If you use `+page.js` it will "restart" the promise. If you use `+page.server.js`, it will stream in the promise result (no restart).

so does that mean you are able to use `defer` in `+page.server.ts` and conditionally activate it if its a data request or SSR?

File: documentation/docs/20-core-concepts/20-load.md
Comment (Rich-Harris) [inline documentation/docs/20-core-concepts/20-load.md:484]:
```suggestion
{#await data.deferred.slow}
```

File: documentation/docs/20-core-concepts/20-load.md
Comment (Rich-Harris) [inline documentation/docs/20-core-concepts/20-load.md:490]:
```suggestion
{/await}
```

File: packages/kit/src/runtime/client/client.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/client/client.js:1844]:
```suggestion
							entry.uses.parent ||= uses.parent;
							entry.uses.route ||= uses.route;
							entry.uses.url ||= uses.url;
```

File: packages/kit/src/runtime/server/data/index.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/data/index.js:214]:
```suggestion
								async ({ d, e }) => {
```

File: packages/kit/src/runtime/server/page/render.js
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/page/render.js:295]: noted in https://github.com/sveltejs/kit/pull/9018#discussion_r1108175556 but we should change one or the other

**Comment by Rich-Harris** [comment]
> If we stream a promise we don't know if all the `uses` fields are final yet. After a promise is resolved someone could access `params`, which only then are marked as used (if not accessed before). So we need to delay (or send again) the `uses` value after all promises of a load are resolved

> I think it would be totally fine if we said that you need to access those things before returning from `load` — I don't think I would have expected it to work differently as a user. Maybe we try shipping without that and see if it turns out to be an issue in practice?

Here's a thing I'm trying to wrap my head around — what happens if the user navigates while data is still streaming in? Is there a situation in which `load` functions would be considered still valid because we didn't yet have a full accounting of `uses`?

If so, I kinda think we should stick to only tracking dependencies before the function returns, so that we can tell people 'if you want to be reactive to `params.foo` changing you need to access it before returning, not inside a promise's `then` callback'. Otherwise the behaviour might be frustratingly non-deterministic

**Comment by paoloricciuti** [comment]
> If so, I kinda think we should stick to only tracking dependencies before the function returns, so that we can tell people 'if you want to be reactive to `params.foo` changing you need to access it before returning, not inside a promise's `then` callback'. Otherwise the behaviour might be frustratingly non-deterministic

As a user I'd say this is totally fine, there a similar constraint for effects in SolidJS and I don't think people are complaining about it.

File: packages/kit/src/runtime/server/data/index.js
Comment (dummdidumm) [inline packages/kit/src/runtime/server/data/index.js:235]: This warning needs a better text - best I could come up with on a friday evening

**Comment by Rich-Harris** [comment]
> As a user I'd say this is totally fine

This is good validation — appreciate you weighing in. Per 1da068e9f2d48f403089f684c0e658834f7e0185 there'll be a warning in the rare cases people get this wrong

File: packages/kit/types/internal.d.ts
Comment (Rich-Harris) [inline packages/kit/types/internal.d.ts:195]: is there a reason this and `ServerNodesResponse` are types rather than interfaces?

**Comment by paoloricciuti** [comment]
@Rich-Harris out of curiosity: I see you are updating the docs to include the fact that it the platform does not support streaming the promise will be buffered...what does it means and how can kit detect if it's on a platform that support streaming? Shouldn't this be handled by adapters?

File: packages/kit/src/runtime/server/data/index.js
Comment (dummdidumm) [inline packages/kit/src/runtime/server/data/index.js:124]: Should've added a comment: I didn't always stream because there's an issue we closed some time ago where we added content length headers to the results, and that goes missing when always streaming.

Comment (Rich-Harris) [reply packages/kit/src/runtime/server/data/index.js:124]: ah, fair enough — was just trying to push everything down a single code path. will revert

**Comment by Rich-Harris** [comment]
@paoloricciuti Lambda gives you an `(event) => { body, ... }` signature, with no access to the standard `Request` and `Response` objects you get almost everywhere else. 'Buffered' means you get the whole response in one go, i.e. at the speed of the slowest promise.

There isn't really a reliable a way for SvelteKit to detect it, and there isn't a way for adapters to know if you use streaming or not, so documentation is the best approach to dealing with this

File: packages/kit/src/runtime/server/data/index.js
Comment (dummdidumm) [reply packages/kit/src/runtime/server/data/index.js:124]: Could we conditionally add a content length header? Then we wouldn't need an extra if path

**Comment by paoloricciuti** [comment]
> @paoloricciuti Lambda gives you an `(event) => { body, ... }` signature, with no access to the standard `Request` and `Response` objects you get almost everywhere else. 'Buffered' means you get the whole response in one go, i.e. at the speed of the slowest promise.
> 
> There isn't really a reliable a way for SvelteKit to detect it, and there isn't a way for adapters to know if you use streaming or not, so documentation is the best approach to dealing with this

Got it, well reading from the commits it seems like everything it's handled automatically so maybe I'd specify that if you deploy in places where streaming it's not supported you should avoid returning nested promises right? Because, correct me if I'm wrong, if kit can't detect if streaming is supported it can't change the type of response so it will still try to stream that promise and it will fail.

**Comment by Rich-Harris** [comment]
It'll still work, it'll just be buffered

**Comment by paoloricciuti** [comment]
> It'll still work, it'll just be buffered

Oh well that's cool than...I tought the request would fail straight away.

P.s. I'm still commenting here also to understand better how it works and to provide some feedback from a user of Sveltekit but I understand it might mean losing time by your part so feel free to shut me down I'll not take it personally 😉

**Comment by Rich-Harris** [comment]
One thing I'm seeing is that the browser doesn't resolve promises one at a time for the initial render, it batches them up. Not sure if there's a minimum chunk size or something — tried adding a newline and it did nothing. Still investigating.

**Comment by Rich-Harris** [comment]
Made a few tweaks that were enabled by not tracking `uses` post-return, notably separating the 'main' payload from subsequent chunks. A few remaining thoughts:

* we probably need to disable etag generation for streamed responses
* we should disable streaming when `csr === false`. error? warning?
* need to figure out why all the `<script>` blocks at the end get buffered, and if there's a solution
* would probably be good to merge #8957 soon so that we can align the `window.whatever` stuff

**Comment by Rich-Harris** [comment]
Okay: `<script type="module">` is deferred until `DOMContentLoaded`. So we need to use regular scripts for the streamed chunks.

This might also mean we need to use a non-module script for the main init block, which seems like a giant nuisance. AFK for a while though, so will investigate more later.

**Comment by Rich-Harris** [comment]
Ok, we can use `import(...)` inside a regular `<script>`, so that's what we'll have to do

**Comment by dummdidumm** [comment]
Is it a problem that it's deferred though? The main module script is deferred, too, so we don't win anything performance-wise when changing it to a regular script

**Comment by Rich-Harris** [comment]
That's what I mean — if we wait for the entire document to load before we hydrate the app, then nothing will be hydrated until the slowest promise resolves. So all the scripts on the page need to be non-module scripts

**Comment by dummdidumm** [comment]
Oooh so that domcontentloaded event is only fired after the last chunk is streamed in? Damn.. yes then we need to change that for all scripts, bummer

**Comment by paoloricciuti** [comment]
Random suggestion but given that we get to choose what to stream what if we append at the end of the actual page (the last piece before the promises to be clear) a script tag that emits a DOMContentLoaded event? Or even a custom event that triggers hydration?

**Comment by Rich-Harris** [comment]
@paoloricciuti you've misunderstood — the script that triggers hydration _doesn't load_ until the DOM is ready. that's not something we can fix with a synthetic/custom event.

I've converted everything to a single simple `<script>` that contains all the setup/loading/starting code, plus the service worker registration. I actually think it's much nicer this way.

For some reason responses are buffered in `vite preview`. Not sure what's going on. Other than that it seems to be working pretty well

**Comment by paoloricciuti** [comment]
> @paoloricciuti you've misunderstood — the script that triggers hydration _doesn't load_ until the DOM is ready. that's not something we can fix with a synthetic/custom event.
> 
> I've converted everything to a single simple `<script>` that contains all the setup/loading/starting code, plus the service worker registration. I actually think it's much nicer this way.
> 
> For some reason responses are buffered in `vite preview`. Not sure what's going on. Other than that it seems to be working pretty well

Duh right...sorry for the misunderstanding

**Comment by Rich-Harris** [comment]
Ah, a fun wrinkle — if you return `params` from `load`, its properties will be accessed during serialization, giving a false positive warning. Not sure how best to solve it. We can't really serialize early because all nodes need to be serialized together to avoid duplication; we can't set a 'you can now warn' flag after serialization because that would introduce race conditions. Might need to get creative with proxies

**Review by Rich-Harris — APPROVED** [review]
I think this is ready to go. Couple of follow-ups I'd like us to do:

1. Figure out why `vite preview` doesn't stream responses. I'm truly baffled by this — the same app running through `adapter-node` works just fine
2. Refactor `render.js` a bit. I started on this in the `refactor-render` branch but scrapped it for this PR as it would have added more noise. But I think there's some good opportunities to make the logic flow a bit more naturally (also the function should probably be renamed `render_html` or something)

File: documentation/docs/20-core-concepts/20-load.md
Comment (geoffrich) [inline documentation/docs/20-core-concepts/20-load.md:423]: I wonder if the title should be something more descriptive so people looking for the equivalent of `+loading.svelte` or whatever can scan for it easily.

File: documentation/docs/20-core-concepts/20-load.md
Comment (geoffrich) [inline documentation/docs/20-core-concepts/20-load.md:425]: I think some additional context as to _why_ you would want to do this would be helpful. Also, do we have guidance for when not to do this? (since it affects the SSR'd page)

```suggestion
Promises at the _top level_ of the returned object will be awaited, making it easy to return multiple promises without creating a waterfall. When using a server `load`, _nested_ promises will be streamed to the browser as they resolve. This is useful if you have slow, non-essential data, since you can start rendering the page before the `fetch` call resolves:
```

**Review by geoffrich — COMMENTED** [review]
Left a few comments on the docs. Also, did we ever make a decision around the option to buffer (or not) when calling `invalidate`?

File: documentation/docs/20-core-concepts/20-load.md
Comment (geoffrich) [reply documentation/docs/20-core-concepts/20-load.md:425]: Copying over some potential guidance from another conversation:

> I wonder if we want to draw a 'essential/non-essential' distinction rather than 'fast/slow' (not least because streaming requires JS)? E.g. your blog post is essential — if [slug] doesn't exist in the database, that's a 404, if it fails to load it's a 500 — but the comments might be non-essential

File: documentation/docs/20-core-concepts/20-load.md
Comment (Rich-Harris) [reply documentation/docs/20-core-concepts/20-load.md:423]: Maybe this?

```suggestion
## Streaming with promises
```

We could add something between the two code blocks like this...

> This is useful for creating skeleton loading states:

...since 'skeleton' and 'loading states' are two things people might search for

File: documentation/docs/20-core-concepts/20-load.md
Comment (Rich-Harris) [reply documentation/docs/20-core-concepts/20-load.md:425]: I like that, though the bit about `fetch` implies that it only applies to client-side nav rather than the initial page render. Maybe this?

```suggestion
Promises at the _top level_ of the returned object will be awaited, making it easy to return multiple promises without creating a waterfall. When using a server `load`, _nested_ promises will be streamed to the browser as they resolve. This is useful if you have slow, non-essential data, since you can start rendering the page before all the data is available:
```

File: documentation/docs/20-core-concepts/20-load.md
Comment (Rich-Harris) [reply documentation/docs/20-core-concepts/20-load.md:472]:
```suggestion
> Streaming data will only work when JavaScript is enabled. You should avoid returning nested promises from a universal `load` function if the page server rendered, as these are _not_ streamed — instead, the promise is recreated when the function re-runs in the browser.
```

**Comment by Rich-Harris** [comment]
Just realised we don't have proper error handling. Will add some

**Comment by Smirow** [comment]
Maybe a stupid remark; I am guessing this should not work in `+layout.server.ts`? (It's not explicit in the docs).

**Comment by Rich-Harris** [comment]
It'll work in `+layout.server.ts` just the same. `await parent()` works the same way it does today — top-level promises are awaited, but other than that it just returns the object returned from the parent `load`

**Pull request merged**
